File: conductor/apps/core/health_check.py
"""
Health Check & Monitoring Service

Production monitoring with:
1. Health check endpoints
2. Dependency status checks
3. Performance metrics
4. System resource monitoring
5. Readiness/Liveness probes
"""

import logging
import time
import os
import platform
from datetime import timedelta
from typing import Dict, Any, List, Optional
from enum import Enum

from django.utils import timezone
from django.db import connection
from django.core.cache import cache
from django.conf import settings

logger = logging.getLogger(__name__)


class HealthStatus(Enum):
    """Health status levels."""
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


class ComponentType(Enum):
    """System components."""
    DATABASE = "database"
    CACHE = "cache"
    CELERY = "celery"
    STORAGE = "storage"
    EMAIL = "email"
    AI_SERVICE = "ai_service"


class HealthCheckService:
    """
    Comprehensive health check and monitoring service.
    """
    
    # Timeout for checks (seconds)
    CHECK_TIMEOUT = 5
    
    # ==========================================================================
    # MAIN HEALTH CHECK
    # ==========================================================================
    
    @classmethod
    def get_health(cls, detailed: bool = False) -> Dict[str, Any]:
        """
        Get overall system health status.
        
        Args:
            detailed: Include detailed component status
            
        Returns:
            Health status response
        """
        start_time = time.time()
        
        # Check all components
        components = cls._check_all_components()
        
        # Determine overall status
        overall_status = cls._calculate_overall_status(components)
        
        response = {
            'status': overall_status.value,
            'timestamp': timezone.now().isoformat(),
            'version': getattr(settings, 'VERSION', '1.0.0'),
            'environment': getattr(settings, 'ENVIRONMENT', 'development'),
            'response_time_ms': round((time.time() - start_time) * 1000, 2)
        }
        
        if detailed:
            response['components'] = components
            response['system'] = cls._get_system_info()
        
        return response
    
    @classmethod
    def _check_all_components(cls) -> Dict[str, Dict]:
        """Check all system components."""
        return {
            ComponentType.DATABASE.value: cls._check_database(),
            ComponentType.CACHE.value: cls._check_cache(),
            ComponentType.CELERY.value: cls._check_celery(),
            ComponentType.STORAGE.value: cls._check_storage(),
        }
    
    @classmethod
    def _calculate_overall_status(cls, components: Dict) -> HealthStatus:
        """Calculate overall status from components."""
        statuses = [c.get('status', 'unhealthy') for c in components.values()]
        
        if all(s == 'healthy' for s in statuses):
            return HealthStatus.HEALTHY
        elif any(s == 'unhealthy' for s in statuses):
            return HealthStatus.UNHEALTHY
        else:
            return HealthStatus.DEGRADED
    
    # ==========================================================================
    # COMPONENT CHECKS
    # ==========================================================================
    
    @classmethod
    def _check_database(cls) -> Dict[str, Any]:
        """Check database connectivity."""
        start = time.time()
        
        try:
            with connection.cursor() as cursor:
                cursor.execute("SELECT 1")
            
            latency = (time.time() - start) * 1000
            
            return {
                'status': 'healthy',
                'latency_ms': round(latency, 2),
                'message': 'Database connection OK'
            }
        except Exception as e:
            return {
                'status': 'unhealthy',
                'latency_ms': -1,
                'message': str(e)
            }
    
    @classmethod
    def _check_cache(cls) -> Dict[str, Any]:
        """Check cache (Redis) connectivity."""
        start = time.time()
        
        try:
            test_key = '_health_check_'
            cache.set(test_key, 'ok', timeout=10)
            result = cache.get(test_key)
            cache.delete(test_key)
            
            latency = (time.time() - start) * 1000
            
            if result == 'ok':
                return {
                    'status': 'healthy',
                    'latency_ms': round(latency, 2),
                    'message': 'Cache connection OK'
                }
            else:
                return {
                    'status': 'degraded',
                    'latency_ms': round(latency, 2),
                    'message': 'Cache read/write mismatch'
                }
        except Exception as e:
            return {
                'status': 'unhealthy',
                'latency_ms': -1,
                'message': str(e)
            }
    
    @classmethod
    def _check_celery(cls) -> Dict[str, Any]:
        """Check Celery worker status."""
        try:
            from celery import current_app
            
            # Inspect active workers
            inspect = current_app.control.inspect(timeout=2)
            stats = inspect.stats()
            
            if stats:
                return {
                    'status': 'healthy',
                    'workers': len(stats),
                    'message': f'{len(stats)} worker(s) active'
                }
            else:
                return {
                    'status': 'degraded',
                    'workers': 0,
                    'message': 'No active workers'
                }
        except Exception as e:
            return {
                'status': 'unhealthy',
                'workers': 0,
                'message': str(e)
            }
    
    @classmethod
    def _check_storage(cls) -> Dict[str, Any]:
        """Check file storage availability."""
        try:
            from django.core.files.storage import default_storage
            
            # Try to list files
            default_storage.listdir('')
            
            return {
                'status': 'healthy',
                'message': 'Storage accessible'
            }
        except Exception as e:
            return {
                'status': 'unhealthy',
                'message': str(e)
            }
    
    # ==========================================================================
    # KUBERNETES PROBES
    # ==========================================================================
    
    @classmethod
    def liveness_probe(cls) -> Dict[str, Any]:
        """
        Kubernetes liveness probe.
        Returns True if the application is running.
        """
        return {
            'status': 'alive',
            'timestamp': timezone.now().isoformat()
        }
    
    @classmethod
    def readiness_probe(cls) -> Dict[str, Any]:
        """
        Kubernetes readiness probe.
        Returns True if the application is ready to serve traffic.
        """
        # Check critical dependencies
        db_check = cls._check_database()
        cache_check = cls._check_cache()
        
        is_ready = (
            db_check.get('status') == 'healthy' and
            cache_check.get('status') in ['healthy', 'degraded']
        )
        
        return {
            'status': 'ready' if is_ready else 'not_ready',
            'database': db_check.get('status'),
            'cache': cache_check.get('status'),
            'timestamp': timezone.now().isoformat()
        }
    
    # ==========================================================================
    # SYSTEM INFO
    # ==========================================================================
    
    @classmethod
    def _get_system_info(cls) -> Dict[str, Any]:
        """Get system information."""
        import psutil
        
        try:
            cpu_percent = psutil.cpu_percent(interval=0.1)
            memory = psutil.virtual_memory()
            disk = psutil.disk_usage('/')
            
            return {
                'platform': platform.system(),
                'python_version': platform.python_version(),
                'cpu_cores': os.cpu_count(),
                'cpu_percent': cpu_percent,
                'memory': {
                    'total_gb': round(memory.total / (1024**3), 2),
                    'used_gb': round(memory.used / (1024**3), 2),
                    'percent': memory.percent
                },
                'disk': {
                    'total_gb': round(disk.total / (1024**3), 2),
                    'used_gb': round(disk.used / (1024**3), 2),
                    'percent': round(disk.percent, 1)
                }
            }
        except ImportError:
            return {
                'platform': platform.system(),
                'python_version': platform.python_version(),
                'cpu_cores': os.cpu_count(),
                'note': 'psutil not available for detailed stats'
            }
        except Exception as e:
            return {'error': str(e)}
    
    # ==========================================================================
    # METRICS
    # ==========================================================================
    
    @classmethod
    def get_metrics(cls) -> Dict[str, Any]:
        """
        Get application metrics for monitoring.
        """
        from apps.users.models import User
        from apps.courses.models import Course, Enrollment
        
        return {
            'users': {
                'total': User.objects.count(),
                'active_today': User.objects.filter(
                    last_login__date=timezone.now().date()
                ).count()
            },
            'courses': {
                'total': Course.objects.filter(is_published=True).count(),
                'enrollments_today': Enrollment.objects.filter(
                    created_at__date=timezone.now().date()
                ).count()
            },
            'system': cls._get_system_info(),
            'timestamp': timezone.now().isoformat()
        }


# ==========================================================================
# API VIEWS
# ==========================================================================

def health_check_view(request):
    """Health check endpoint."""
    from django.http import JsonResponse
    
    detailed = request.GET.get('detailed', 'false').lower() == 'true'
    health = HealthCheckService.get_health(detailed=detailed)
    
    status_code = 200 if health['status'] == 'healthy' else 503
    
    return JsonResponse(health, status=status_code)


def liveness_view(request):
    """Kubernetes liveness probe endpoint."""
    from django.http import JsonResponse
    
    return JsonResponse(HealthCheckService.liveness_probe())


def readiness_view(request):
    """Kubernetes readiness probe endpoint."""
    from django.http import JsonResponse
    
    result = HealthCheckService.readiness_probe()
    status_code = 200 if result['status'] == 'ready' else 503
    
    return JsonResponse(result, status=status_code)


def metrics_view(request):
    """Metrics endpoint for monitoring."""
    from django.http import JsonResponse
    from rest_framework.permissions import IsAdminUser
    
    # In production, add authentication
    metrics = HealthCheckService.get_metrics()
    
    return JsonResponse(metrics)

File: learninghub/backend/src/monitoring/healthCheck.ts
import { Request, Response } from 'express'
import { prisma } from '../config'

const startTime = Date.now()

interface HealthStatus {
  status: 'healthy' | 'degraded' | 'unhealthy'
  version: string
  uptime: number
  uptimeHuman: string
  timestamp: string
  environment: string
  checks: {
    database: { status: string; latencyMs: number; error?: string }
    memory: {
      status: string
      heapUsedMB: number
      heapTotalMB: number
      rssMB: number
      percentUsed: number
    }
    system: { status: string; nodeVersion: string; platform: string; cpuUsage: NodeJS.CpuUsage }
  }
}

function formatUptime(ms: number): string {
  const seconds = Math.floor(ms / 1000)
  const minutes = Math.floor(seconds / 60)
  const hours = Math.floor(minutes / 60)
  const days = Math.floor(hours / 24)

  if (days > 0) return `${days}d ${hours % 24}h ${minutes % 60}m`
  if (hours > 0) return `${hours}h ${minutes % 60}m ${seconds % 60}s`
  if (minutes > 0) return `${minutes}m ${seconds % 60}s`
  return `${seconds}s`
}

/**
 * @swagger
 * /health:
 *   get:
 *     summary: Comprehensive health check endpoint
 *     description: Reports database connectivity, redis connectivity, memory usage, system info, and uptime.
 *     tags: [Monitoring]
 *     responses:
 *       200:
 *         description: System is completely healthy
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 status:
 *                   type: string
 *                   example: "healthy"
 *                 uptimeHuman:
 *                   type: string
 *       503:
 *         description: System is degraded or unhealthy
 */
export async function healthCheck(_req: Request, res: Response): Promise<void> {
  const uptimeMs = Date.now() - startTime
  const memUsage = process.memoryUsage()
  const heapUsedMB = Math.round((memUsage.heapUsed / 1024 / 1024) * 100) / 100
  const heapTotalMB = Math.round((memUsage.heapTotal / 1024 / 1024) * 100) / 100
  const rssMB = Math.round((memUsage.rss / 1024 / 1024) * 100) / 100
  const percentUsed = Math.round((memUsage.heapUsed / memUsage.heapTotal) * 100)

  // Database check
  let dbStatus = 'ok'
  let dbLatency = 0
  let dbError: string | undefined

  try {
    const dbStart = Date.now()
    await prisma.$queryRaw`SELECT 1`
    dbLatency = Date.now() - dbStart
    if (dbLatency > 1000) {
      dbStatus = 'slow'
    }
  } catch (err) {
    dbStatus = 'error'
    dbError = err instanceof Error ? err.message : 'Unknown database error'
  }

  // Redis check
  const { cacheService } = await import('../services/CacheService')
  let redisStatus = 'ok'
  const redisEnabled = process.env.REDIS_ENABLED === 'true'
  if (redisEnabled) {
    const isRedisHealthy = await cacheService.healthCheck()
    if (!isRedisHealthy) {
      redisStatus = 'error'
    }
  } else {
    redisStatus = 'disabled'
  }

  // Memory check
  const memoryStatus = percentUsed > 90 ? 'critical' : percentUsed > 75 ? 'warning' : 'ok'

  // Overall status
  let overallStatus: 'healthy' | 'degraded' | 'unhealthy' = 'healthy'
  if (dbStatus === 'error') overallStatus = 'unhealthy'
  else if (dbStatus === 'slow' || memoryStatus === 'warning' || redisStatus === 'error')
    overallStatus = 'degraded'
  else if (memoryStatus === 'critical') overallStatus = 'unhealthy'

  const health: HealthStatus & { checks: { redis: { status: string } } } = {
    status: overallStatus,
    version: process.env.npm_package_version ?? '1.0.0',
    uptime: uptimeMs,
    uptimeHuman: formatUptime(uptimeMs),
    timestamp: new Date().toISOString(),
    environment: process.env.NODE_ENV ?? 'development',
    checks: {
      database: {
        status: dbStatus,
        latencyMs: dbLatency,
        ...(dbError && { error: dbError }),
      },
      redis: {
        status: redisStatus,
      },
      memory: {
        status: memoryStatus,
        heapUsedMB,
        heapTotalMB,
        rssMB,
        percentUsed,
      },
      system: {
        status: 'ok',
        nodeVersion: process.version,
        platform: process.platform,
        cpuUsage: process.cpuUsage(),
      },
    },
  }

  const statusCode = overallStatus === 'unhealthy' ? 503 : 200
  res.status(statusCode).json(health)
}

/**
 * @swagger
 * /health/live:
 *   get:
 *     summary: Lightweight liveness probe
 *     description: Returns 200 if the process is alive. No dependency checks.
 *     tags: [Monitoring]
 *     responses:
 *       200:
 *         description: Service is alive
 */
export function livenessProbe(_req: Request, res: Response): void {
  res.status(200).json({ status: 'alive', timestamp: new Date().toISOString() })
}

/**
 * Readiness probe — checks if the service can handle requests.
 * Verifies database connectivity.
 *
 * GET /api/v1/health/ready
 */
export async function readinessProbe(_req: Request, res: Response): Promise<void> {
  try {
    await prisma.$queryRaw`SELECT 1`

    let redisReady = true
    if (process.env.REDIS_ENABLED === 'true') {
      const { cacheService } = await import('../services/CacheService')
      redisReady = await cacheService.healthCheck()
    }

    if (!redisReady) {
      res.status(503).json({
        status: 'not_ready',
        reason: 'redis_unavailable',
        timestamp: new Date().toISOString(),
      })
      return
    }

    res.status(200).json({ status: 'ready', timestamp: new Date().toISOString() })
  } catch {
    res.status(503).json({
      status: 'not_ready',
      reason: 'database_unavailable',
      timestamp: new Date().toISOString(),
    })
  }
}
